File: src/oidc.ts
import * as crypto from 'crypto';

export interface OIDCConfig {
  issuer: string;
  authorization_endpoint: string;
  token_endpoint: string;
  router_url?: string;
}

export interface OIDCDeps {
  fetch?: (url: string) => Promise<{ ok: boolean; status: number; json: () => Promise<any> }>;
}

/**
 * Generate a PKCE code_verifier (43-128 chars, unreserved characters per RFC 7636)
 */
export function generateCodeVerifier(length: number = 64): string {
  if (length < 43 || length > 128) {
    throw new Error('code_verifier length must be between 43 and 128');
  }
  const unreserved = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~';
  const bytes = crypto.randomBytes(length);
  let result = '';
  for (let i = 0; i < length; i++) {
    result += unreserved[bytes[i] % unreserved.length];
  }
  return result;
}

/**
 * Generate PKCE code_challenge using S256: BASE64URL(SHA256(code_verifier))
 */
export function generateCodeChallenge(verifier: string): string {
  const hash = crypto.createHash('sha256').update(verifier, 'ascii').digest();
  return hash.toString('base64url');
}

/**
 * Generate a random state string for CSRF prevention
 */
export function generateState(): string {
  return crypto.randomBytes(32).toString('base64url');
}

/**
 * Discover OIDC endpoints from tenant domain
 */
export async function discoverOIDC(
  tenantDomain: string,
  deps?: OIDCDeps
): Promise<OIDCConfig> {
  const fetchFn = deps?.fetch ?? globalThis.fetch;
  const scheme = tenantDomain.startsWith('localhost') || tenantDomain.startsWith('127.0.0.1') ? 'http' : 'https';
  const url = `${scheme}://${tenantDomain}/.well-known/openid-configuration`;

  let response: any;
  try {
    response = await fetchFn(url);
  } catch (err: any) {
    throw new Error(`Failed to connect to OIDC provider at ${tenantDomain}: ${err.message}`);
  }

  if (!response.ok) {
    throw new Error(`OIDC Discovery failed (HTTP ${response.status}) for ${tenantDomain}`);
  }

  const config = await response.json();

  if (!config.authorization_endpoint || !config.token_endpoint || !config.issuer) {
    throw new Error(`Invalid OIDC Discovery response from ${tenantDomain}: missing required fields`);
  }

  return {
    issuer: config.issuer,
    authorization_endpoint: config.authorization_endpoint,
    token_endpoint: config.token_endpoint,
    router_url: config.router_url,
  };
}

File: src/oidc.rs
//! OIDC discovery, PKCE generation, and tenant→Stark domain resolution.

use base64::Engine;
use rand::RngCore;
use serde::Deserialize;
use sha2::{Digest, Sha256};

use crate::error::{AppError, Result};
use crate::net::Client;

const UNRESERVED: &[u8] = b"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~";

#[derive(Debug, Clone)]
pub struct OIDCConfig {
    pub issuer: String,
    pub authorization_endpoint: String,
    pub token_endpoint: String,
    pub device_authorization_endpoint: Option<String>,
    pub router_url: Option<String>,
}

/// PKCE `code_verifier` (43-128 chars, unreserved characters per RFC 7636).
pub fn generate_code_verifier(length: usize) -> Result<String> {
    if !(43..=128).contains(&length) {
        return Err(AppError::new(
            "code_verifier length must be between 43 and 128",
        ));
    }
    let mut bytes = vec![0u8; length];
    rand::thread_rng().fill_bytes(&mut bytes);
    let s: String = bytes
        .iter()
        .map(|b| UNRESERVED[(*b as usize) % UNRESERVED.len()] as char)
        .collect();
    Ok(s)
}

/// PKCE `code_challenge` using S256: BASE64URL(SHA256(code_verifier)).
pub fn generate_code_challenge(verifier: &str) -> String {
    let digest = Sha256::digest(verifier.as_bytes());
    base64::engine::general_purpose::URL_SAFE_NO_PAD.encode(digest)
}

/// Random `state` for CSRF prevention (32 bytes, base64url).
pub fn generate_state() -> String {
    let mut bytes = [0u8; 32];
    rand::thread_rng().fill_bytes(&mut bytes);
    base64::engine::general_purpose::URL_SAFE_NO_PAD.encode(bytes)
}

/// Default Stark (control plane) domains by environment.
pub fn stark_domain_for_env(env: &str) -> &'static str {
    match env {
        "prod" => "monocle-ai.com",
        "stg" => "stg.monocle-ai.com",
        "local" => "localhost:9000",
        _ => "monocle-ai.com",
    }
}

/// Resolve the Stark domain from a tenant domain. Tenants always run on a
/// subdomain — bare domains are rejected.
///
///   stg-warmblood091803.monocle-ai.com → stg.monocle-ai.com
///   warmblood.monocle-ai.com           → monocle-ai.com
///   localhost:8080                      → localhost:8080
pub fn resolve_stark_domain(tenant_domain: &str) -> Result<String> {
    if tenant_domain.starts_with("localhost") || tenant_domain.starts_with("127.0.0.1") {
        return Ok(tenant_domain.to_string());
    }

    let parts: Vec<&str> = tenant_domain.split('.').collect();
    if parts.len() <= 2 {
        return Err(AppError::new(format!(
            "Invalid tenant domain: {tenant_domain}. Tenant must be a subdomain (e.g., mytenant.monocle-ai.com)"
        )));
    }

    let subdomain = parts[0];
    let base_domain = parts[1..].join(".");

    if subdomain.starts_with("stg-") {
        Ok(format!("stg.{base_domain}"))
    } else {
        Ok(base_domain)
    }
}

fn scheme_for(domain: &str) -> &'static str {
    if domain.starts_with("localhost") || domain.starts_with("127.0.0.1") {
        "http"
    } else {
        "https"
    }
}

#[derive(Deserialize)]
struct DiscoveryDoc {
    issuer: Option<String>,
    authorization_endpoint: Option<String>,
    token_endpoint: Option<String>,
    device_authorization_endpoint: Option<String>,
    router_url: Option<String>,
}

/// Discover OIDC endpoints from a Stark domain.
pub fn discover_oidc(client: &Client, domain: &str) -> Result<OIDCConfig> {
    let url = format!(
        "{}://{}/.well-known/openid-configuration",
        scheme_for(domain),
        domain
    );

    let resp = client.get(&url, &[]).map_err(|e| {
        AppError::new(format!(
            "Failed to connect to OIDC provider at {domain}: {e}"
        ))
    })?;

    if !resp.ok() {
        return Err(AppError::new(format!(
            "OIDC Discovery failed (HTTP {}) for {domain}",
            resp.status
        )));
    }

    let doc: DiscoveryDoc = resp.json()?;

    match (doc.authorization_endpoint, doc.token_endpoint, doc.issuer) {
        (Some(authorization_endpoint), Some(token_endpoint), Some(issuer)) => Ok(OIDCConfig {
            issuer,
            authorization_endpoint,
            token_endpoint,
            device_authorization_endpoint: doc.device_authorization_endpoint,
            router_url: doc.router_url,
        }),
        _ => Err(AppError::new(format!(
            "Invalid OIDC Discovery response from {domain}: missing required fields"
        ))),
    }
}
